File: frontend/src/lib/risk.ts
import type { Risk, RiskCategory, RiskLevel } from "@/types/risk";

export interface RiskMatrixSnapshot {
  probability: number;
  impact: number;
  cellKey: string;
  probabilityLabel: string;
  impactLabel: string;
}

export interface RiskScoreSnapshot {
  probability: number;
  impact: number;
  weight: number;
  nilai: number;
  score: number;
  level: RiskLevel;
  priority: number;
  matrix: RiskMatrixSnapshot;
}

export interface ResolvedRiskScoreSemantics {
  source: "inherent";
  usesReviewed: false;
  isFinalized: boolean;
  effective: RiskScoreSnapshot;
  primary: RiskScoreSnapshot;
  inherent: RiskScoreSnapshot;
}

type RiskScoreBundleInput = {
  probability: number;
  impact: number;
  weight: number;
  nilai?: number | null;
  score?: number | null;
};

type RiskScoreSemanticFields = Pick<
  Risk,
  | "status"
  | "probability"
  | "impact"
  | "weight"
  | "nilai"
  | "inherentScore"
>;

export const PROBABILITY_LABELS: Record<number, string> = {
  1: "Jarang",
  2: "Kemungkinan Kecil",
  3: "Kemungkinan Sedang",
  4: "Kemungkinan Besar",
  5: "Hampir Pasti Terjadi",
};

export const IMPACT_LABELS: Record<number, string> = {
  1: "Tidak Signifikan",
  2: "Kecil",
  3: "Sedang",
  4: "Besar",
  5: "Katastropik",
};

export const riskCategoryLabels: Record<RiskCategory, string> = {
  "": "Belum dikategorikan",
  kebijakan: "Kebijakan",
  reputasi: "Reputasi",
  fraud_korupsi: "Fraud - Korupsi",
  legal: "Legal",
  kepatuhan: "Kepatuhan",
  operasional: "Operasional",
};

export const dashboardCategoryLabels: Record<string, string> = {
  kebijakan: "Kebijakan",
  reputasi: "Reputasi",
  fraud_korupsi: "Fraud - Korupsi",
  legal: "Legal",
  kepatuhan: "Kepatuhan",
  operasional: "Operasional",
  uncategorized: "Tanpa Kategori",
};

// Bobot Matrix 5x5 based on Probability (rows) and Impact (columns)
// Rows: Probability 1-5 (Jarang to Hampir Pasti Terjadi)
// Cols: Impact 1-5 (Tdk Signifikan to Katastropik)
export const BobotMatrix: number[][] = [
  // Impact: 1(Tdk Signifikan), 2(Kecil), 3(Sedang), 4(Besar), 5(Katastropik)
  [1.0, 1.5, 2.0, 3.0, 4.0],      // Prob 1: Jarang
  [1.0, 1.8, 1.83, 1.9, 2.1],     // Prob 2: Kemungkinan Kecil
  [1.17, 1.42, 1.43, 1.46, 1.47], // Prob 3: Kemungkinan Sedang
  [1.2, 1.19, 1.3, 1.16, 1.2],    // Prob 4: Kemungkinan Besar
  [1.5, 1.4, 1.13, 1.15, 1.0],    // Prob 5: Hampir Pasti Terjadi
];

// Get bobot from matrix based on probability and impact (1-5)
export function getBobot(probability: number, impact: number): number {
  if (probability < 1 || probability > 5 || impact < 1 || impact > 5) {
    return 1.0;
  }
  return BobotMatrix[probability - 1][impact - 1];
}

// Calculate nilai = probability × impact × weight
export function calculateNilai(probability: number, impact: number, weight: number): number {
  return Math.round(probability * impact * weight * 100) / 100;
}

// Get risk level based on nilai (new Indonesian levels)
// Sangat Rendah: < 5, Rendah: 5-9, Sedang: 10-14, Tinggi: 15-19, Sangat Tinggi: >= 20
export function getRiskLevelFromNilai(nilai: number): RiskLevel {
  const rounded = Math.round(nilai);
  if (rounded >= 20) return "sangat_tinggi";
  if (rounded >= 15) return "tinggi";
  if (rounded >= 10) return "sedang";
  if (rounded >= 5) return "rendah";
  return "sangat_rendah";
}

// Legacy function - kept for backward compatibility
export function getRiskLevel(score: number): RiskLevel {
  if (score >= 20) return "sangat_tinggi";
  if (score >= 15) return "tinggi";
  if (score >= 10) return "sedang";
  if (score >= 5) return "rendah";
  return "sangat_rendah";
}

// Get risk priority based on risk level
// Sangat Tinggi = 1, Tinggi = 2, Sedang = 3, Rendah = 4, Sangat Rendah = 5
export function getRiskPriority(level: RiskLevel): number {
  const priorities: Record<RiskLevel, number> = {
    sangat_tinggi: 1,
    tinggi: 2,
    sedang: 3,
    rendah: 4,
    sangat_rendah: 5,
  };
  return priorities[level] || 5;
}

export interface RiskAssessmentClassification {
  score: number;
  level: RiskLevel;
  priority: number;
  appetite: "dalam_batas" | "di_atas_batas";
  isRiskUtama: boolean;
}

export function resolveRiskAssessmentClassification(
  nilai: number,
): RiskAssessmentClassification {
  const score = Math.round(nilai);
  const level = getRiskLevelFromNilai(nilai);

  return {
    score,
    level,
    priority: getRiskPriority(level),
    appetite: resolveRiskAppetite(score),
    isRiskUtama: isRiskUtama(score),
  };
}

export function getRiskLevelLabel(level: RiskLevel): string {
  const labels: Record<RiskLevel, string> = {
    sangat_rendah: "Sangat Rendah",
    rendah: "Rendah",
    sedang: "Sedang",
    tinggi: "Tinggi",
    sangat_tinggi: "Sangat Tinggi",
  };
  return labels[level] || level;
}

export function levelToColor(level: RiskLevel): string {
  const colors: Record<RiskLevel, string> = {
    sangat_rendah: "bg-green-100 text-green-700 border-green-200",
    rendah: "bg-risk-low/15 text-risk-low border-risk-low/20",
    sedang: "bg-risk-medium/15 text-risk-medium border-risk-medium/20",
    tinggi: "bg-risk-high/15 text-risk-high border-risk-high/20",
    sangat_tinggi: "bg-risk-extreme/15 text-risk-extreme border-risk-extreme/20",
  };
  return colors[level] || "";
}

export function levelToFillColor(level: RiskLevel): string {
  const colors: Record<RiskLevel, string> = {
    sangat_rendah: "rgb(74 222 128)",
    rendah: "var(--color-risk-low)",
    sedang: "var(--color-risk-medium)",
    tinggi: "var(--color-risk-high)",
    sangat_tinggi: "var(--color-risk-extreme)",
  };
  return colors[level] || "var(--color-risk-low)";
}

export function getRiskLevelDisplayLabel(level: RiskLevel): string {
  const labels: Record<RiskLevel, string> = {
    sangat_tinggi: "Sangat Tinggi",
    tinggi: "Tinggi",
    sedang: "Sedang",
    rendah: "Rendah",
    sangat_rendah: "Sangat Rendah",
  };
  return labels[level] || level;
}

export function getLevelBadgeClasses(level: string): string {
  switch (level) {
    case "Ekstrem":
    case "Sangat Tinggi":
    case "sangat_tinggi":
      return "bg-red-500/15 text-red-700 border-red-500/30";
    case "Tinggi":
    case "tinggi":
      return "bg-orange-500/15 text-orange-700 border-orange-500/30";
    case "Sedang":
    case "sedang":
      return "bg-yellow-500/15 text-yellow-700 border-yellow-500/30";
    case "Rendah":
    case "rendah":
      return "bg-blue-500/15 text-blue-700 border-blue-500/30";
    default:
      return "bg-emerald-500/15 text-emerald-700 border-emerald-500/30";
  }
}

export function getScoreBtnColorClasses(score: number): string {
  if (score >= 20) return "bg-red-500 text-white hover:bg-red-600";
  if (score >= 15) return "bg-orange-500 text-white hover:bg-orange-600";
  if (score >= 10) return "bg-yellow-500 text-white hover:bg-yellow-600";
  if (score >= 5) return "bg-blue-500 text-white hover:bg-blue-600";
  return "bg-emerald-500 text-white hover:bg-emerald-600";
}

// Calculate all risk metrics at once
export function calculateRiskMetrics(probability: number, impact: number) {
  const weight = getBobot(probability, impact);
  const nilai = calculateNilai(probability, impact, weight);
  const inherentScore = Math.round(nilai);
  const level = getRiskLevelFromNilai(nilai);
  const priority = getRiskPriority(level);

  return {
    weight,
    nilai,
    inherentScore,
    level,
    priority,
  };
}

/**
 * Resolve risk appetite status based on inherentScore.
 * Per KMK: inherentScore < 10 → "dalam_batas", >= 10 → "di_atas_batas"
 */
export function resolveRiskAppetite(inherentScore: number): "dalam_batas" | "di_atas_batas" {
	return inherentScore < 10 ? "dalam_batas" : "di_atas_batas";
}

/**
 * Returns true if risk is "risk utama" (Sedang or higher).
 * Per KMK: inherentScore >= 10 corresponds to Sedang/Tinggi/SangatTingkat.
 */
export function isRiskUtama(inherentScore: number): boolean {
	return inherentScore >= 10;
}

function isExplicitNumber(value: number | null | undefined): value is number {
  return value !== null && value !== undefined;
}

function buildRiskMatrixSnapshot(probability: number, impact: number): RiskMatrixSnapshot {
  return {
    probability,
    impact,
    cellKey: `${probability}-${impact}`,
    probabilityLabel: PROBABILITY_LABELS[probability] ?? String(probability),
    impactLabel: IMPACT_LABELS[impact] ?? String(impact),
  };
}

function buildRiskScoreSnapshot(bundle: RiskScoreBundleInput): RiskScoreSnapshot {
  const nilai = isExplicitNumber(bundle.nilai)
    ? bundle.nilai
    : calculateNilai(bundle.probability, bundle.impact, bundle.weight);
  const score = isExplicitNumber(bundle.score) ? bundle.score : Math.round(nilai);
  const level = getRiskLevelFromNilai(nilai);

  return {
    probability: bundle.probability,
    impact: bundle.impact,
    weight: bundle.weight,
    nilai,
    score,
    level,
    priority: getRiskPriority(level),
    matrix: buildRiskMatrixSnapshot(bundle.probability, bundle.impact),
  };
}

export function resolveRiskScoreSemantics(risk: RiskScoreSemanticFields): ResolvedRiskScoreSemantics {
  const inherent = buildRiskScoreSnapshot({
    probability: risk.probability,
    impact: risk.impact,
    weight: risk.weight,
    nilai: risk.nilai,
    score: risk.inherentScore,
  });

  return {
    source: "inherent",
    usesReviewed: false,
    isFinalized: risk.status === "approved",
    effective: inherent,
    primary: inherent,
    inherent,
  };
}

// Simpulan utilities for Risk Assessment
export function getSimpulanTingkatRisiko(nilaiCurrent: number, nilaiBaru: number): string {
  if (nilaiBaru === nilaiCurrent) return "Tidak ada penurunan tingkat risiko";
  if (nilaiBaru > nilaiCurrent) return "Tingkat risiko mengalami peningkatan";
  return "Tingkat risiko mengalami penurunan";
}

export function getSimpulanEfektifitas(nilaiCurrent: number, nilaiBaru: number): string {
  if (nilaiBaru < nilaiCurrent) return "Efektif";
  return "Tidak Efektif";
}

export function getSimpulanTingkatRisikoColor(nilaiCurrent: number, nilaiBaru: number): string {
  if (nilaiBaru === nilaiCurrent) return "bg-yellow-100 text-yellow-700 border-yellow-200";
  if (nilaiBaru > nilaiCurrent) return "bg-red-100 text-red-700 border-red-200";
  return "bg-green-100 text-green-700 border-green-200";
}

export function getSimpulanEfektifitasColor(nilaiCurrent: number, nilaiBaru: number): string {
  if (nilaiBaru < nilaiCurrent) return "bg-green-100 text-green-700 border-green-200";
  return "bg-red-100 text-red-700 border-red-200";
}

File: backend/internal/domain/entity/risk.go
package entity

import (
	"math"
	"time"

	"github.com/google/uuid"
	"github.com/manris/backend/internal/domain/errors"
)

// RiskLevel constants for Indonesian risk levels
const (
	RiskLevelSangatRendah = "sangat_rendah"
	RiskLevelRendah       = "rendah"
	RiskLevelSedang       = "sedang"
	RiskLevelTinggi       = "tinggi"
	RiskLevelSangatTinggi = "sangat_tinggi"
)

// RiskStatus constants for risk workflow states
const (
	RiskStatusDraft    = "assessment_draft"
	RiskStatusInReview = "assessment_in_review"
	RiskStatusApproved = "approved"
)

// BobotMatrix is the 5x5 weight matrix based on Probability (rows) and Impact (columns)
// Rows: Probability 1-5 (Jarang to Hampir Pasti Terjadi)
// Cols: Impact 1-5 (Tdk Signifikan to Katastropik)
var BobotMatrix = [5][5]float64{
	// Impact: 1(Tdk Signifikan), 2(Kecil), 3(Sedang), 4(Besar), 5(Katastropik)
	{1.0, 1.5, 2.0, 3.0, 4.0},      // Prob 1: Jarang
	{1.0, 1.8, 1.83, 1.9, 2.1},     // Prob 2: Kemungkinan Kecil
	{1.17, 1.42, 1.43, 1.46, 1.47}, // Prob 3: Kemungkinan Sedang
	{1.2, 1.19, 1.3, 1.16, 1.2},    // Prob 4: Kemungkinan Besar
	{1.5, 1.4, 1.13, 1.15, 1.0},    // Prob 5: Hampir Pasti Terjadi
}

// Risk represents a risk register entry
type Risk struct {
	ID                     uuid.UUID  `json:"id"`
	Code                   string     `json:"code"`
	Title                  string     `json:"title"`
	Description            string     `json:"description"`
	Category               string     `json:"category"`
	Status                 string     `json:"status"`
	VersionGroupID         uuid.UUID  `json:"versionGroupId"`
	PreviousRiskID         *uuid.UUID `json:"previousRiskId,omitempty"`
	IsCurrent              bool       `json:"isCurrent"`
	IsCycleCurrent         bool       `json:"isCycleCurrent"`
	VersionNumber          int        `json:"versionNumber"`
	ArchivedAt             *time.Time `json:"archivedAt,omitempty"`
	ArchivedReason         string     `json:"archivedReason,omitempty"`
	OrganizationID         *uuid.UUID `json:"organizationId,omitempty"`
	OrgName                string     `json:"orgName"`
	CreatedBy              *uuid.UUID `json:"createdBy,omitempty"`
	CreatedByName          string     `json:"createdByName"`
	RiskOwnerID            *uuid.UUID `json:"riskOwnerId,omitempty"`
	ControlOwnerID         *uuid.UUID `json:"controlOwnerId,omitempty"`
	ObjectiveID            *uuid.UUID `json:"objectiveId,omitempty"`
	ROID                   *uuid.UUID `json:"roId,omitempty"`
	LikelihoodAssessmentID *uuid.UUID `json:"likelihoodAssessmentId,omitempty"`
	ImpactCriteriaID       *uuid.UUID `json:"impactCriteriaId,omitempty"`
	ImpactJustification    string     `json:"impactJustification,omitempty"`

	// Section 1
	Cause           []string `json:"cause,omitempty"`
	RiskSource      string   `json:"riskSource,omitempty"`
	Controllability string   `json:"controllability,omitempty"`
	ImpactDesc      []string `json:"impactDesc,omitempty"`

	// Section 2
	ExistingControl      string  `json:"existingControl,omitempty"`
	ControlEffectiveness string  `json:"controlEffectiveness,omitempty"`
	Probability          int     `json:"probability"`
	Impact               int     `json:"impact"`
	Weight               float64 `json:"weight,omitempty"`
	Nilai                float64 `json:"nilai,omitempty"` // Nilai = Probability × Impact × Weight
	InherentScore        int     `json:"inherentScore"`

	// Section 3
	RiskPriority    int    `json:"riskPriority,omitempty"`
	RiskAppetite    string `json:"riskAppetite,omitempty"`
	TreatmentOption string `json:"treatmentOption,omitempty"`

	// Section 4
	Mitigations []Mitigation `json:"mitigations,omitempty"`

	// Section 5
	TargetProbability        int                  `json:"targetProbability,omitempty"`
	TargetImpact             int                  `json:"targetImpact,omitempty"`
	TargetWeight             float64              `json:"targetWeight,omitempty"`
	TargetNilai              float64              `json:"targetNilai,omitempty"`
	TargetScore              int                  `json:"targetScore,omitempty"`
	ResidualAcceptanceReason string               `json:"residualAcceptanceReason,omitempty"`
	NextReviewDate           *string              `json:"nextReviewDate,omitempty"`
	ReviewScheduleText       string               `json:"reviewScheduleText,omitempty"`
	AssessmentCycle          string               `json:"assessmentCycle,omitempty"`
	ReviewType               string               `json:"reviewType,omitempty"`
	ChangeReason             string               `json:"changeReason,omitempty"`
	ReviewSummary            string               `json:"reviewSummary,omitempty"`
	ReviewStartedAt          *time.Time           `json:"reviewStartedAt,omitempty"`
	ReviewSubmittedAt        *time.Time           `json:"reviewSubmittedAt,omitempty"`
	ReviewApprovedAt         *time.Time           `json:"reviewApprovedAt,omitempty"`
	DraftApprovalLine        []ApprovalLineMember `json:"draftApprovalLine,omitempty"`

	// Ongoing draft tracking (for list views)
	DraftID               *uuid.UUID                `json:"draftId,omitempty"`
	DraftStatus           *string                   `json:"draftStatus,omitempty"`
	HasOngoing            bool                      `json:"hasOngoing"`
	MonitoringStatus      *string                   `json:"monitoringStatus,omitempty"`
	LastMonitoredAt       *time.Time                `json:"lastMonitoredAt,omitempty"`
	BeforeMonitoringNilai *float64                  `json:"beforeMonitoringNilai,omitempty"`
	MonitoringResultNilai *float64                  `json:"monitoringResultNilai,omitempty"`
	SemesterMonitoring    *SemesterMonitoringStatus `json:"semesterMonitoring,omitempty"`

	CreatedAt time.Time `json:"createdAt"`
	UpdatedAt time.Time `json:"updatedAt"`
}

type SemesterMonitoringStatus struct {
	H1 *string `json:"h1"`
	H2 *string `json:"h2"`
}

type ApprovalLineMember struct {
	ID   string `json:"id"`
	Name string `json:"name"`
	Type string `json:"type,omitempty"` // 'review' or 'approval' - distinguishes reviewer from approver
}

const (
	RiskCategoryKebijakan   = "kebijakan"
	RiskCategoryOperasional = "operasional"
	RiskCategoryKepatuhan   = "kepatuhan"
	RiskCategoryFraud       = "fraud_korupsi"
	RiskCategoryReputasi    = "reputasi"
	RiskCategoryLegal       = "legal"
)

var allowedRiskCategories = map[string]struct{}{
	"":                      {},
	RiskCategoryKebijakan:   {},
	RiskCategoryOperasional: {},
	RiskCategoryKepatuhan:   {},
	RiskCategoryFraud:       {},
	RiskCategoryReputasi:    {},
	RiskCategoryLegal:       {},
}

func IsValidRiskCategory(category string) bool {
	_, ok := allowedRiskCategories[category]
	return ok
}

// Validate performs domain validation on Risk
func (r *Risk) Validate() error {
	if r.Title == "" {
		return errors.ErrInvalidTitle
	}
	if r.Code == "" {
		return errors.ErrInvalidCode
	}
	if r.Status == "" {
		return errors.ErrInvalidStatus
	}
	if !IsValidRiskCategory(r.Category) {
		return errors.ErrInvalidRiskCategory
	}
	if r.Probability < 1 || r.Probability > 5 {
		return errors.ErrInvalidProbability
	}
	if r.Impact < 1 || r.Impact > 5 {
		return errors.ErrInvalidImpact
	}
	for i := range r.Mitigations {
		if err := r.Mitigations[i].Validate(); err != nil {
			return err
		}
	}
	return nil
}

// CanBeSubmittedForApproval checks if risk can be submitted for approval
func (r *Risk) CanBeSubmittedForApproval() bool {
	return r.Status == RiskStatusDraft
}

// GetBobot returns the weight from the matrix based on probability and impact
func GetBobot(probability, impact int) float64 {
	if probability < 1 || probability > 5 || impact < 1 || impact > 5 {
		return 1.0
	}
	return BobotMatrix[probability-1][impact-1]
}

// CalculateNilai calculates nilai = probability × impact × weight
func CalculateNilai(probability, impact int, weight float64) float64 {
	raw := float64(probability) * float64(impact) * weight
	return math.Round(raw*100) / 100
}

// GetInherentScore calculates inherent score
func (r *Risk) GetInherentScore() int {
	return int(math.Round(float64(r.Probability) * float64(r.Impact) * r.Weight))
}

// CalculateInherentScore calculates and sets the inherent score
func (r *Risk) CalculateInherentScore() {
	r.InherentScore = int(math.Round(float64(r.Probability) * float64(r.Impact) * r.Weight))
}

// CalculateTargetScore calculates and sets the target score
func (r *Risk) CalculateTargetScore() {
	r.TargetScore = int(math.Round(float64(r.TargetProbability) * float64(r.TargetImpact) * r.TargetWeight))
}

// CalculateBobot calculates and sets the weight based on probability and impact
func (r *Risk) CalculateBobot() {
	r.Weight = GetBobot(r.Probability, r.Impact)
}

// CalculateNilai calculates and sets the nilai field
func (r *Risk) CalculateNilai() {
	r.Nilai = CalculateNilai(r.Probability, r.Impact, r.Weight)
}

// CalculateTargetBobot calculates and sets the target weight based on target probability and impact
func (r *Risk) CalculateTargetBobot() {
	r.TargetWeight = GetBobot(r.TargetProbability, r.TargetImpact)
}

// CalculateTargetNilai calculates and sets the target nilai field
func (r *Risk) CalculateTargetNilai() {
	r.TargetNilai = CalculateNilai(r.TargetProbability, r.TargetImpact, r.TargetWeight)
}

// GetRiskLevel returns Indonesian risk level based on nilai
// Sangat Rendah: nilai < 5, Rendah: 5-9, Sedang: 10-14, Tinggi: 15-19, Sangat Tinggi: >= 20
func (r *Risk) GetRiskLevel() string {
	return GetRiskLevelFromNilai(r.EffectiveNilai())
}

// GetRiskLevelFromNilai returns Indonesian risk level based on nilai value
func GetRiskLevelFromNilai(nilai float64) string {
	rounded := math.Round(nilai)
	switch {
	case rounded >= 20:
		return RiskLevelSangatTinggi
	case rounded >= 15:
		return RiskLevelTinggi
	case rounded >= 10:
		return RiskLevelSedang
	case rounded >= 5:
		return RiskLevelRendah
	default:
		return RiskLevelSangatRendah
	}
}

// GetRiskPriority returns priority based on risk level
// Sangat Tinggi = 1, Tinggi = 2, Sedang = 3, Rendah = 4, Sangat Rendah = 5
func (r *Risk) GetRiskPriority() int {
	return GetRiskPriorityFromLevel(r.GetRiskLevel())
}

// GetRiskPriorityFromLevel returns priority number based on risk level string
func GetRiskPriorityFromLevel(level string) int {
	switch level {
	case RiskLevelSangatTinggi:
		return 1
	case RiskLevelTinggi:
		return 2
	case RiskLevelSedang:
		return 3
	case RiskLevelRendah:
		return 4
	case RiskLevelSangatRendah:
		return 5
	default:
		return 5
	}
}

// GetRiskLevelDisplay returns the Indonesian display name for risk level
func GetRiskLevelDisplay(level string) string {
	switch level {
	case RiskLevelSangatTinggi:
		return "Sangat Tinggi"
	case RiskLevelTinggi:
		return "Tinggi"
	case RiskLevelSedang:
		return "Sedang"
	case RiskLevelRendah:
		return "Rendah"
	case RiskLevelSangatRendah:
		return "Sangat Rendah"
	default:
		return level
	}
}

// ResolveRiskAppetite returns advisory appetite status based on inherentScore.
// Per KMK risk appetite matrix: inherentScore < 10 → dalam_batas, >= 10 → di_atas_batas.
func ResolveRiskAppetite(inherentScore int) string {
	if inherentScore < 10 {
		return "dalam_batas"
	}
	return "di_atas_batas"
}

// IsRiskUtama returns true if risk level is Sedang or higher.
// Per KMK: inherentScore >= 10 corresponds to Sedang/Tinggi/SangatTinggi level.
func (r Risk) IsRiskUtama() bool {
	return r.InherentScore >= 10
}

// GetRiskAppetiteDisplay returns the Indonesian display name for risk appetite
func GetRiskAppetiteDisplay(appetite string) string {
	switch appetite {
	case "dalam_batas":
		return "Dalam batas selera risiko"
	case "di_atas_batas":
		return "Di atas batas selera risiko"
	default:
		return appetite
	}
}

// GetTreatmentOptionDisplay returns the Indonesian display name for treatment option
func GetTreatmentOptionDisplay(option string) string {
	switch option {
	case "avoid", "menghindari":
		return "Menghindari Risiko"
	case "transfer", "berbagi":
		return "Berbagi Risiko"
	case "mitigate", "mitigasi":
		return "Mitigasi"
	case "accept", "menerima":
		return "Menerima Risiko"
	default:
		return option
	}
}

// GetControlEffectivenessDisplay returns the Indonesian display name for control effectiveness
func GetControlEffectivenessDisplay(eff string) string {
	switch eff {
	case "efektif":
		return "Efektif"
	case "tidak_efektif":
		return "Tidak Efektif"
	default:
		return eff
	}
}

// GetControllabilityDisplay returns the Indonesian display name for controllability
func GetControllabilityDisplay(c string) string {
	switch c {
	case "C":
		return "Controllable"
	case "UC":
		return "Uncontrollable"
	default:
		return c
	}
}

// GetRiskSourceDisplay returns the Indonesian display name for risk source
func GetRiskSourceDisplay(s string) string {
	switch s {
	case "internal":
		return "Internal"
	case "eksternal":
		return "Eksternal"
	default:
		return s
	}
}

// CalculateAll computes bobot, nilai, inherent score, and updates risk priority
func (r *Risk) CalculateAll() {
	r.CalculateBobot()
	r.CalculateNilai()
	r.CalculateInherentScore()
	r.RiskPriority = r.GetRiskPriority()
}

// IsLocked checks if risk is in a locked (non-editable) status
func (r *Risk) IsLocked() bool {
	return r.Status == RiskStatusInReview || r.Status == RiskStatusApproved
}

// IsApprovedCurrent returns whether this risk is the active approved version.
func (r *Risk) IsApprovedCurrent() bool {
	return r.Status == RiskStatusApproved && r.IsCurrent
}

// CanBeReassessed returns whether the risk can start a periodic reassessment.
func (r *Risk) CanBeReassessed() bool {
	return r.IsApprovedCurrent()
}

// AddMitigation adds a mitigation to the risk
func (r *Risk) AddMitigation(mitigation Mitigation) {
	r.Mitigations = append(r.Mitigations, mitigation)
}

func (r *Risk) effectivePreliminaryScore() int {
	if r.InherentScore > 0 {
		return r.InherentScore
	}
	if r.Nilai > 0 {
		return int(math.Round(r.Nilai))
	}
	weight := r.Weight
	if weight == 0 {
		weight = GetBobot(r.Probability, r.Impact)
	}
	return int(math.Round(float64(r.Probability) * float64(r.Impact) * weight))
}

func (r *Risk) EffectiveProbability() int {
	return r.Probability
}

func (r *Risk) EffectiveImpact() int {
	return r.Impact
}

func (r *Risk) EffectiveNilai() float64 {
	return r.Nilai
}

func (r *Risk) GetEffectiveScore() int {
	return r.effectivePreliminaryScore()
}
